Overhaul sequence parsing and sprite loading

pchote opened the pull request:
This PR follows on from #20580, completely overhauling the code implementation for sequences to remove longstanding technical debt and add plumbing to support several new features for the remastered assets.

I strongly suggest reviewing by commit.

Previously, everything to do with sequence parsing and sprite loading was done from the `ISpriteSequence` implementation's constructor. This forced awkward workarounds using `Lazy<>` to defer parsing and error handling so the utility could run in contexts where the assets are not available. It also led to code smells like having to call virtual methods from the base constructor.

This PR introduces a new approach where the sequence parsing and sprite loading are made distinct. The old `SequenceProvider` has been renamed to `SequenceSet` to match its parallel role to the `Ruleset`, and the `SpriteCache` has been completely rewritten to minimise the memory overhead when loading frames (no need to buffer frames) and provide a central place for handling missing files.

This also removes the `ModData.DefaultSequences` cache, choosing deliberately to reload everything from scratch during map load. This does, in theory, increase the load times between the lobby and gameplay, but in practice sprite loading is now fast enough for the difference to be negligible in mods that have a reasonable amount of low-resolution sprites. Also, many players (certainly in the RA MP community) play on custom maps that bypass that optimisation anyway. Mods that have high-resolution or an unreasonable amount of sprites will see noticeably longer lobby->gameplay load times, but significantly reduced RAM/VRAM usage as its no longer necessary to hold multiple copies of the artwork in memory. This is an important optimisation for TD remastered where each copy of the sequence artwork is ~1GB!

`DefaultSpriteSequence` has been rearranged to be able to support new ways of specifying filenames and to allow pre-processing (e.g. masking) of frame data before it is added to the sprite sheet. See [RemasterSpriteSequence.cs](https://github.com/pchote/TiberianDawnHD/blob/master/OpenRA.Mods.Mobius/RemasterSpriteSequence.cs) for testcases - particularly the `RemasteredFilenamePattern` and `RemasteredMaskFilename` implementations.

pchote committed:
Add ILintSequencesPass.

pchote committed:
Access sequences from Map.

pchote committed:
Remove asset browser dependency on sequences.

pchote committed:
Remove internal state from ISpriteSequence API.

pchote committed:
Fix negative sequence lengths.

pchote committed:
Allow all sprites to use interpolated facings.

AspectInteractive2 commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -84,16 +84,20 @@ public static WAngle AngleDiffToStep(WAngle facing, int numFrames)
+		/// <summary>Returns the angle that the closest facing sprite should be rotated by to achieve the closest interpolated facing.</summary>
+		public static WAngle GetInterpolatedFacingRotation(WAngle facing, int facings, int interpolatedFacings)
 		{
+			if (facings == interpolatedFacings)
+				return WAngle.Zero;
```
Currently it is impossible for `facings` to be = to `interpolatedFacings` due to the below code in `DefaultSpriteSequence.cs`. I see you have kept this check in this overhaul as well, so we could save performance by omitting this if statement.

``` C#
				if (interpolatedFacings != -1 && (interpolatedFacings <= 1 || interpolatedFacings <= Math.Abs(facings) || interpolatedFacings > 1024
					|| !Exts.IsPowerOf2(interpolatedFacings)))
					throw new YamlException($"InterpolatedFacings must be greater than Facings, within the range of 2 to 1024, and a power of 2.");
```

AspectInteractive2 commented on OpenRA.Mods.Common/Traits/QuantizeFacingsFromSequence.cs:
```diff
@@ -23,14 +23,13 @@ public class QuantizeFacingsFromSequenceInfo : ConditionalTraitInfo, IQuantizeBo
 			var rsi = ai.TraitInfo<RenderSpritesInfo>();
-			var seq = sequenceProvider.GetSequence(rsi.GetImage(ai, race), Sequence);
-			return seq.InterpolatedFacings == -1 ? seq.Facings : seq.InterpolatedFacings;
+			return sequences.GetSequence(rsi.GetImage(ai, faction), Sequence).Facings;
```
This is how it was previously and it did not work because `seq.InterpolatedFacings` needs to be returned if `InterpolatedFacings` are being used, not `Facings`. This is because `Facings` is never overridden by `InterpolatedFacings` since the interpolation is purely visual. so if you return `Facings` in all cases you are going to get the non-interpolated facings for all the sprites.

AspectInteractive2 replied:
This seems to have not changed in the latest commit.

pchote replied:
Interpolated vs "real" facings is an internal implementation detail that users don't need to know about!

`.Facings` returns the interpolated value if it is defined.

AspectInteractive2 replied:
Ok have tested and seems to still work fine. I do agree, though I remember making the change because it didn't work otherwise, but since it's working looks to be okay.

pchote commented:
Argh, seems I had made a mess of the interpolated facings feature during a previous rebase. Updated.

penev92 commented on OpenRA.Mods.Common/SpriteLoaders/PngSheetLoader.cs:
```diff
@@ -31,6 +30,26 @@ public PngSheetMetadata(Dictionary<string, string> metadata)
 		}
 	}
 
+	public class EmbeddedSpritePalette
```
As D2k is still *the* main user of this it makes no sense for it to be in `PngSheetLoader`. Can we give it its own file please?

pchote replied:
It's part of the png(sheet) format so IMO it does make sense here, but putting it in its own file is probably fine too.

penev92 replied:
`PngSheetLoader` and `R8Loader` both use it identically in their `TryParseSprite` methods and it (and `PaletteFromEmbeddedSpritePalette`) is (now) generic enough that it doesn't need to be tied to a specific format.
Also class-per-file is a very basic convention that is just a good idea.
Would be nice to give it its own file.

penev92 commented on OpenRA.Game/Graphics/Animation.cs:
```diff
@@ -22,7 +22,7 @@ public class Animation
 		public bool IsDecoration { get; set; }
 
-		readonly SequenceProvider sequenceProvider;
+		readonly SequenceProvider sequences;
```
I personally dislike the `sequenceProvider` -> `sequences` rename when it's for a `SequenceProvider` instance.
I get that it may have been for consistency because we currently use both names for it, but I do not enjoy the direction this took.

penev92 replied:
I see - preparation for later `SequenceProvider` rename/rework.

penev92 commented on OpenRA.Game/Graphics/SequenceProvider.cs:
```diff
@@ -22,16 +22,12 @@ namespace OpenRA.Graphics
 		int ZOffset { get; }
-		int ShadowStart { get; }
 		int ShadowZOffset { get; }
-		int[] Frames { get; }
```
What is the criteria for what gets removed and what stays here?

pchote replied:
Internal implementation details about how sprites are laid out within their source files are removed, things that are relevant to their usage by the game stay.

penev92 commented on OpenRA.Mods.Common/UtilityCommands/DumpSequenceSheetsCommand.cs:
```diff
@@ -34,17 +34,19 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-			SequenceProvider sequences;
-			if (!modData.DefaultSequences.TryGetValue(args[2], out sequences))
+			SequenceSet sequences;
+			if (!modData.DefaultTerrainInfo.ContainsKey(args[2]))
```
Can you please invert the `if`?

penev92 commented on OpenRA.Game/Map/MapPreview.cs:
```diff
@@ -226,7 +226,7 @@ public Ruleset LoadRuleset()
 			return Ruleset.Load(modData, this, TileSet, innerData.RuleDefinitions,
 				innerData.WeaponDefinitions, innerData.VoiceDefinitions, innerData.NotificationDefinitions,
-				innerData.MusicDefinitions, innerData.SequenceDefinitions, innerData.ModelSequenceDefinitions);
+				innerData.MusicDefinitions, innerData.ModelSequenceDefinitions);
```
I'm not sure I'm comfortable with how special sprite sequences are becoming by being the only exception here.

pchote replied:
Sprite sequences (and terrain info) were already an exception (sometimes null, sometimes not, parsed in a completely different way)! The changes here are cleaning up those hacks.

pchote commented:
Rebased and updated.

pchote commented:
Added a commit to fix the SequenceSet disposal.

pchote commented:
Rebased and ready for review.

penev92 commented on OpenRA.Mods.Common/Traits/Palettes/PaletteFromEmbeddedSpritePalette.cs:
```diff
@@ -27,13 +26,11 @@ public class PaletteFromEmbeddedSpritePaletteInfo : TraitInfo, IProvidesCursorPa
 		[FieldLoader.Require]
-		[Desc("Sequence image holding the palette definition")]
-		public readonly string Image = null;
+		[Desc("Sprite filename image holding the palette definition.")]
```
"Sprite filename image" doesn't sound right.

penev92 commented on OpenRA.Mods.Common/TraitsInterfaces.cs:
```diff
@@ -35,7 +35,7 @@ public enum ResupplyType
 	public interface IQuantizeBodyOrientationInfo : ITraitInfoInterface
 	{
-		int QuantizedBodyFacings(ActorInfo ai, SequenceProvider sequenceProvider, string race);
+		int QuantizedBodyFacings(ActorInfo ai, SequenceSet sequences, string race);
```
Please use this opportunity to change the last remaining occurrence of "race" to "faction", like you already did in `QuantizeFacingsFromSequence.QuantizedBodyFacings()`.

penev92 commented on OpenRA.Game/GameRules/Ruleset.cs:
```diff
@@ -146,7 +142,7 @@ public static Ruleset LoadDefaults(ModData modData)
 
 				// The default ruleset does not include a preferred tileset or sequence set
-				ruleset = new Ruleset(actors, weapons, voices, notifications, music, null, null, modelSequences);
+				ruleset = new Ruleset(actors, weapons, voices, notifications, music, null, modelSequences);
```
With the removal of the SequenceSet parameter, should the above comment be updated?

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -21,49 +21,50 @@ namespace OpenRA.Mods.Common.Graphics
 					{
-						var sequence = CreateSequence(modData, tileSet, cache, imageNode.Key, sequenceNode.Key, sequenceNode.Value, defaults);
+						var sequence = CreateSequence(modData, tileset, cache, imageNode.Key, sequenceNode.Key, sequenceNode.Value, defaults);
+						((DefaultSpriteSequence)sequence).ReserveSprites(modData, tileset, cache, sequenceNode.Value, defaults);
```
Why is this not part of the `ISpriteSequence` interface?

pchote replied:
It's an internal implementation detail that the gameplay code shouldn't have visibility of.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -260,247 +268,260 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
+			return usedFrames.ToArray();
+		}
+
+		protected virtual IEnumerable<(string Filename, int[] loadFrames, int[] useFrames, MiniYamlNode.SourceLocation location)> ParseFilenames(ModData modData, string tileset, int[] frames, MiniYaml data, MiniYaml defaults)
```
A bit of inconsistent naming - `Filename` vs `loadFrames`, `useFrames`, `location`. They should all start with a capital letter.
Also `useFrames` sounds like a boolean.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -260,247 +268,260 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
+		{
+			var filename = LoadField(Filename, data, defaults, out var location);
+
+			// Only request the subset of frames that we actually need
```
Fullstop.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -260,247 +268,260 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
+			yield return (filename, loadFrames, frames, location);
+		}
+
+		protected virtual IEnumerable<(string Filename, int[] loadFrames, int[] useFrames, MiniYamlNode.SourceLocation location)> ParseCombineFilenames(ModData modData, string tileset, int[] frames, MiniYaml data)
```
Same thing with naming - capital letters.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -260,247 +268,260 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
+			var depthSprite = LoadField(DepthSprite, data, defaults, out var depthSpriteLocation);
+			if (!string.IsNullOrEmpty(depthSprite))
+				depthSpriteReservation = cache.ReserveSprites(depthSprite, new[] { LoadField(DepthSpriteFrame, data, defaults) }, depthSpriteLocation);
+			depthSpriteOffset = LoadField(DepthSpriteOffset, data, defaults);
```
Empty line above, please.

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -260,247 +268,260 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
 		}
 
-		public DefaultSpriteSequence(ModData modData, string tileSet, SpriteCache cache, ISpriteSequenceLoader loader, string image, string sequence, MiniYaml data, MiniYaml defaults)
+		protected class SpriteReservation
```
Class declaration in the middle of another class is weird. I believe the norm for nested classes is to define them at the very top?

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -295,293 +268,293 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
 
-					return new Sprite(s.Sheet, bounds, zRamp, new float3(dx, dy, dz), s.Channel, blendMode);
-				}).ToArray();
+					foreach ((var subFilename, var subLoadFrames, var subUseFrames, var subLocation) in ParseCombineFilenames(modData, tileset, subFrames, subData))
```
This can be
```suggestion
					foreach (var (subFilename, subLoadFrames, subUseFrames, subLocation) in ParseCombineFilenames(modData, tileset, subFrames, subData))
```

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -295,293 +268,293 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
-					alpha = Exts.MakeArray(length, _ => alpha[0]);
-				else if (alpha.Length != length)
-					throw new YamlException($"Sequence {image}.{sequence} must define either 1 or {length} Alpha values.");
+				foreach ((var filename, var loadFrames, var useFrames, var location) in ParseFilenames(modData, tileset, frames, data, defaults))
```
And this can be
```suggestion
				foreach (var (filename, loadFrames, useFrames, location) in ParseFilenames(modData, tileset, frames, data, defaults))
```

penev92 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -295,293 +268,293 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
-				var ds = cache[depthSprite, GetDepthFrame][depthSpriteFrame];
+				var resolved = cache.ResolveSprites(r.Token);
+				if (r.Frames != null)
+					resolved = r.Frames.Select(f => resolved[f]).ToArray();
```
Is there a reason for `SpriteCache.ResolveSprites()` to return `Sprite[]` instead of `IEnumerable<Sprite>`? We could skip the `ToArray()` here if that is changed.

pchote replied:
Yes, because it is indexed by `f => resolved[f]` which isn't possible with an enumerable.

penev92 commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+
+		public SpriteCache(IReadOnlyFileSystem fileSystem, ISpriteLoader[] loaders, int bgraSheetSize, int indexedSheetSize, int bgraSheetMargin = 1, int indexedSheetMargin = 1)
+		{
+			SheetBuilders = new Dictionary<SheetType, SheetBuilder>()
```
Redundant `()`.

penev92 commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+
+		public int ReserveSprites(string filename, IEnumerable<int> frames, MiniYamlNode.SourceLocation location)
+		{
+			var token = ++nextReservationToken;
```
Not that it matters, but won't the first token always be 2?

penev92 commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+			this.loaders = loaders;
+		}
+
+		public int ReserveSprites(string filename, IEnumerable<int> frames, MiniYamlNode.SourceLocation location)
```
Change parameter to `int[] frames` and skip the `.ToArray()` below? (All usages pass explicit arrays anyway.)
Same for `ReserveFrames()` below.

pchote replied:
The `ToArray`s are actually there to guarantee that other code can't modify the contents of the array in the meantime!

penev92 commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+		{
+			var token = ++nextReservationToken;
+			spriteReservations[token] = (frames?.ToArray(), location);
+			reservationsByFilename.GetOrAdd(filename, new List<int>()).Add(token);
```
This can be simplified to
```suggestion
			reservationsByFilename.GetOrAdd(filename, new List<int> { token });
```
Same for `ReserveFrames()` below.

pchote replied:
This stops it from adding tokens if `filename` is already there!

penev92 commented on OpenRA.Mods.Common/UtilityCommands/CheckMissingSprites.cs:
```diff
@@ -38,29 +38,21 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-						foreach (var image in kv.Value.Images)
+						var sequences = new SequenceSet(modData.DefaultFileSystem, modData, kv.Key, null);
+						sequences.SpriteCache.LoadReservations(modData);
+						foreach ((var filename, var location) in sequences.SpriteCache.MissingFiles)
```
This can be
```suggestion
						foreach (var (filename, location) in sequences.SpriteCache.MissingFiles)
```

penev92 commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+			var spriteCache = new Dictionary<int, Sprite>();
+			foreach (var filename in reservationsByFilename.Keys)
+			{
+				modData.LoadScreen?.Display();
```
It should not be the `SpriteCache`'s job to know about, let alone control the load screen!
Not to mention this is called from two utility commands that don't care about load screens.

pchote replied:
This is needed to prevent the loadscreen from stalling for many seconds, and is consistent with the calls from `VoxelModelSequenceLoader`, `ChromeProvider`, etc. If you can suggest a better place to put this then I can change it!

The `?.` ensures that this is a no-op when run from the utility.

penev92 commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+				sb.Current.CreateBuffer();
+
+			var spriteCache = new Dictionary<int, Sprite>();
+			foreach (var filename in reservationsByFilename.Keys)
```
If you change this to 
```suggestion
			foreach (var (filename, reservations) in reservationsByFilename)
```
you can do `foreach (var token in reservations)` for the below `foreach`, which I personally find preferable.

penev92 commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+			foreach (var filename in reservationsByFilename.Keys)
+			{
+				modData.LoadScreen?.Display();
+				var loadedFrames = GetFrames(fileSystem, filename, loaders, out var fileMetadata);
```
`fileMetadata` is unused! If that is intentional then please switch to 
```suggestion
				var loadedFrames = GetFrames(fileSystem, filename, loaders, out _);
```

penev92 commented on OpenRA.Mods.Cnc/Graphics/ClassicTilesetSpecificSpriteSequence.cs:
```diff
@@ -34,20 +34,56 @@ public class ClassicTilesetSpecificSpriteSequence : ClassicSpriteSequence
+			: base(cache, loader, image, sequence, data, defaults) { }
 
-		protected override string GetSpriteFilename(ModData modData, string tileset, string image, string sequence, MiniYaml data, MiniYaml defaults)
+		protected override IEnumerable<(string Filename, int[] loadFrames, int[] useFrames, MiniYamlNode.SourceLocation location)> ParseFilenames(ModData modData, string tileset, int[] frames, MiniYaml data, MiniYaml defaults)
```
I'm sure I mentioned it in another file, but inconsistent naming - `loadFrames`, `useFrames` and `location` should be PascalCase like `Filename`.

penev92 reviewed (dismissed):
I already did a quick pass review before the prerequisite PR was merged, but here is a more detailed one.
This time there is at least one major concern - the LoadScreen.

AspectInteractive2 commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -1407,5 +1408,10 @@ public string Translate(string key, IDictionary<string, object> args = null)
 			return modData.Translation.GetString(key, args);
 		}
+
+		public void Dispose()
+		{
+			Sequences.Dispose();
+		}
```
This feels like it shouldn't belong in the `Map.cs` class given that it is purely related to sequences (understand there is scope to add more to the Dispose function, but that isn't the case currently). It will need to be moved out of `Map.cs` anyway once we start using [specialised Map formats](https://github.com/OpenRA/OpenRA/pull/20582).

pchote replied:
`Map` owns `Sequences`, so it is its job to dispose it!

AspectInteractive2 commented on OpenRA.Game/ModData.cs:
```diff
@@ -167,6 +158,7 @@ public Map PrepareMap(string uid)
 
 			// Reinitialize all our assets
```
Shouldn't this read 'Initialize' rather than 'Reinitialize'? As both `InitializeLoaders()` and `LoadSprites()` are not called anywhere else in the code base.

pchote replied:
No, as the comment is referring to `InitializeLoaders` which messes with static state that is reinitialized each time a new map is loaded.

AspectInteractive2 commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -84,16 +84,17 @@ public static WAngle AngleDiffToStep(WAngle facing, int numFrames)
 
-		public static WAngle GetInterpolatedFacing(WAngle facing, int facings, int interpolatedFacings)
+		/// <summary>Returns the angle that the closest facing sprite should be rotated by to achieve the closest interpolated facing.</summary>
+		public static WAngle GetInterpolatedFacingRotation(WAngle facing, int facings, int interpolatedFacings)
```
I don't think the name should be changed to `Get...Rotation` unless it is in fact returning a `WRot`. You could call it `Get...Angle` if you wanted to specify the returned object type in the method name.

pchote replied:
Rotation is used everwhere else in the call chain (`Sequence.GetSpriteWithRotation`, `Animation.Render`, `SpriteRenderable`). IMO its clear from the context that this is a 2D rotation of a sprite in screen-space, however, if there is a consensus that the naming should be changed it should be done in its own PR (this one is already too big!).

AspectInteractive2 commented on OpenRA.Mods.Common/UtilityCommands/CheckMissingSprites.cs:
```diff
@@ -38,29 +38,21 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-				// DefaultSequences is a dictionary of tileset: SequenceProvider
-				// so we can also use this to key our tileset checks
-				foreach (var kv in modData.DefaultSequences)
+				foreach (var kv in modData.DefaultTerrainInfo)
```
Further to other feedback, I think this should use the deconstructing version for better readability: `foreach (var (terrainInfoKey, terrainInfo) in modData.DefaultTerrainInfo)`

AspectInteractive2 commented on OpenRA.Mods.Common/UtilityCommands/CheckMissingSprites.cs:
```diff
@@ -38,29 +38,21 @@ void IUtilityCommand.Run(Utility utility, string[] args)
-
-						if (terrainInfo is ITemplatedTerrainInfo templatedTerrainInfo)
+						if (kv.Value is ITemplatedTerrainInfo templatedTerrainInfo)
 							foreach (var r in modData.DefaultRules.Actors[SystemActors.World].TraitInfos<ITiledTerrainRendererInfo>())
```
Can `var r` be changed to `var rule` for clarity? I'm guessing that is what it stands for.

AspectInteractive2 commented on OpenRA.Mods.Common/UtilityCommands/CheckYaml.cs:
```diff
@@ -62,6 +63,13 @@ void IUtilityCommand.Run(Utility utility, string[] args)
 
 					// Run all rule checks on the default mod rules.
 					CheckRules(modData, modData.DefaultRules);
+					foreach (var tileset in modData.DefaultTerrainInfo.Keys)
```
Genuine question: Will these always hold?

Rules being `DefaultRules`
TerrainInfo being `DefaultTerrainInfo`
FileSystem being `DefaultFileSystem`

It seems a bit contrived, as though we are implying extensibility when there isn't any. Maybe a `TO DO: Provide support for multiple / non-standard TerrainInfos here` or similar is in order? Otherwise you may as well remove the 'Default' keyword if no alternatives are planned.

pchote replied:
TerrainInfo can't (yet...?) be modified by maps. I agree, but this request has nothing to do with this PR!

AspectInteractive2 replied:
Well it does directly refer to the code you are adding, even if it isn't the focus of the PR. So I think adding a 'TO DO:' would not be out of line, but your call.

AspectInteractive2 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -139,149 +88,173 @@ public SpriteSequenceField(string key, T defaultValue)
-		protected int facings;
+		protected static readonly SpriteSequenceField<int> Facings = new SpriteSequenceField<int>(nameof(Facings), 1);
 
 		[Desc("The amount of directions the unit faces. Use negative values to rotate counter-clockwise.")]
```
As per #20707 this should be updated to the below (note this is different to the above Facings description which also needs tweaking):

```C#
[Desc("The amount of directions the unit faces after interpolation. Value must be positive, greater than Facings, and a power of 2. ")]
```

AspectInteractive2 commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -139,149 +88,173 @@ public SpriteSequenceField(string key, T defaultValue)
-		int stride;
+		protected static readonly SpriteSequenceField<int> Stride = new SpriteSequenceField<int>(nameof(Stride), -1);
 
 		[Desc("The amount of directions the unit faces. Use negative values to rotate counter-clockwise.")]
```
As per #20707 this should be updated to the below:

```C#
[Desc("The amount of directions the unit faces without interpolation. Use negative values to rotate counter-clockwise.")]
```

pchote committed:
Remove HasEmbeddedPalette from sequences.

pchote commented:
I believe that i've addressed all the comments now in one way or another.

PunkPun commented:
Immediately needs a rebase due to changes to update path

pchote commented:
Done.

PunkPun commented on OpenRA.Mods.Common/Util.cs:
```diff
@@ -84,16 +84,17 @@ public static WAngle AngleDiffToStep(WAngle facing, int numFrames)
 		}
 
-		public static WAngle GetInterpolatedFacing(WAngle facing, int facings, int interpolatedFacings)
+		/// <summary>Returns the angle that the closest facing sprite should be rotated by to achieve the closest interpolated facing.</summary>
```
```suggestion
		/// <returns> The angle that the closest facing sprite should be rotated by to achieve the closest interpolated facing.</returns>
```

We should use this more often

RoosterDragon replied:
`summary` would by my preference here as it's the element you always have. I wouldn't usually employ `returns` unless there was something particularly interesting to say about it not already covered in the summary.

RoosterDragon commented on OpenRA.Game/Graphics/SequenceSet.cs:
```diff
@@ -0,0 +1,119 @@
+		IReadOnlyDictionary<string, ISpriteSequence> ParseSequences(ModData modData, string tileSet, SpriteCache cache, MiniYamlNode node);
+	}
+
+	public class SequenceSet : IDisposable
```
Mark as `sealed` so we don't have to implement the long-winded dispose pattern.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+		{
+			var token = nextReservationToken++;
+			spriteReservations[token] = (frames?.ToArray(), location);
+			reservationsByFilename.GetOrAdd(filename, new List<int>()).Add(token);
```
Use `() => new List<int>()` to avoid allocating empty lists we throw away if there's already an entry in the dictionary. Same in the other reserve method.

RoosterDragon commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,175 @@
+						{
+							var reservation = new ISpriteFrame[loadedFrames.Length];
+							for (var i = 0; i < loadedFrames.Length; i++)
+								if (r.Frames == null || r.Frames.Contains(i))
```
`r.Frames.Contains` inside the loop is going to be quadratic. Are we sure frames will always be small enough to not worry about creating a set outside the loop? Same a bit further down.

RoosterDragon commented on OpenRA.Game/Map/Map.cs:
```diff
@@ -1416,5 +1417,10 @@ public string Translate(string key, IDictionary<string, object> args = null)
 			return modData.Translation.GetString(key, args);
 		}
+
+		public void Dispose()
```
There's one other non-commandy call to a `Map` constructor in `NewMapLogic.cs` that lacks a call to `Dispose`, can we add that?

RoosterDragon commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -294,293 +278,283 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
+			return usedFrames.ToArray();
+		}
+
+		protected virtual IEnumerable<(string Filename, int[] LoadFrames, int[] Frames, MiniYamlNode.SourceLocation Location)> ParseFilenames(ModData modData, string tileset, int[] frames, MiniYaml data, MiniYaml defaults)
```
This 4-tuple occurs in enough places I'd rather see struct for it.

RoosterDragon commented on OpenRA.Mods.Common/Graphics/TilesetSpecificSpriteSequence.cs:
```diff
@@ -32,20 +32,56 @@ public class TilesetSpecificSpriteSequence : DefaultSpriteSequence
 				if (tilesetNode != null)
-					return tilesetNode.Value.Value;
+				{
+					// Only request the subset of frames that we actually need
```
These tileset sequence implementations for ParseFilenames/ParseCombineFilenames occur twice. Is it worth having it shared somewhere? Maybe a protected helper in the base class that impls can choose to employ in this override?

RoosterDragon commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -19,70 +19,51 @@
 					{
-						var sequence = CreateSequence(modData, tileSet, cache, imageNode.Key, sequenceNode.Key, sequenceNode.Value, defaults);
+						var sequence = CreateSequence(modData, tileset, cache, imageNode.Key, sequenceNode.Key, sequenceNode.Value, defaults);
+						((DefaultSpriteSequence)sequence).ReserveSprites(modData, tileset, cache, sequenceNode.Value, defaults);
```
I'm confused - how is this a safe cast if `CreateSequence` is virtual?

pchote replied:
This is safe by policy: users that subclass the `*Loader` are expected to also subclass the same `*SpriteSequence` much like traits/traitinfos.

RoosterDragon commented on OpenRA.Mods.Common/Graphics/DefaultSpriteSequence.cs:
```diff
@@ -294,293 +278,283 @@ protected static Rectangle FlipRectangle(Rectangle rect, bool flipX, bool flipY)
-				// Overrides Length with a custom stride
-				stride = LoadField(Stride.Key, length, data, defaults);
+			if (length == null && facings > 1)
+				throw new YamlException($"{facingsLocation}: {Facings.Key} cannot be used with {Length.Key}: *.");
```
Should `length == null && facings > 1` use `length != null` given the message of `{Facings.Key} cannot be used with {Length.Key}`?

pchote replied:
`length = null` is used to represent "Length: *", i.e. load all the frames, so the current check is correct.

pchote committed:
Dispose SequenceSet when we're done with it.

Utility rules that do something on a map and exit
are left without explicit disposing, as they will
be cleaned up immediately anyway.

PunkPun commented on OpenRA.Mods.Common/UtilityCommands/DumpSequenceSheetsCommand.cs:
```diff
@@ -50,7 +55,9 @@ void IUtilityCommand.Run(Utility utility, string[] args)
 				var max = s == sb.Current ? (int)sb.CurrentChannel + 1 : 4;
 				for (var i = 0; i < max; i++)
-					s.AsPng((TextureChannel)ChannelMasks[i], palette).Save($"{count++}.png");
+					s.AsPng((TextureChannel)ChannelMasks[i], palette).Save($"{count}.{i}.png");
```
```suggestion
					s.AsPng((TextureChannel)ChannelMasks[i], palette).Save($"{count++}.{i}.png");
```

Without this it feels weird. It logs `Saved [0..2].png` but it's actually many more images

<img width="242" alt="Screenshot 2023-03-08 at 12 14 50" src="https://user-images.githubusercontent.com/37534529/223685849-ed33d16d-112f-410b-8a1c-adba41827dba.png">

What exactly is the idea behind this change?

PunkPun replied:
If this is intended. Then perhaps another entry mentioning how many images?

pchote commented:
Updated.

pchote commented:
(airport hopping: will deal with the remaining comment and lint failures later)

pchote committed:
Fix --dump-sequence-sheets utility command.

Also formats output filenames to specify
indexed sprite channels.

pchote committed:
Reduce duplication and allocations around CalculateFrameIndices.

pchote commented:
Updated.

PunkPun reviewed (dismissed):
LGTM

RoosterDragon commented on OpenRA.Game/Graphics/SpriteCache.cs:
```diff
@@ -0,0 +1,174 @@
+		{
+			var token = nextReservationToken++;
+			frameReservations[token] = (frames?.ToArray(), location);
+			reservationsByFilename.GetOrAdd(filename, new List<int>()).Add(token);
```
`_ => new List<int>()` here too please.

pchote committed:
Rewrite sequence loading logic.

Multiple layers of Lazy<T>ness are replaced with
an explicit two-part loading scheme.

Sequences are parsed immediately, without the need
for the sprite assets, and tell the SpriteCache
which frames they need. Use-cases that want the
actual sprites can then tell the SpriteCache to
load the frames and the sequences to resolve the
sprites.

pchote commented:
Fixed.

PunkPun commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/65a8ea73b1ab400913d73605ce04ce2738a5f6e8...0808b1102ef8d23378e6d6e531750ec30bc1fb7c)

Edit by @penev92:
[Chagelog update](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/aa7b7b643107ad002e7ac5a804c46543e10c81e0...6f38f69a20ff10ab87b71a37863b5fa7d56d8db1)
